### PR TITLE
feat(any): extend any to hold enum types via heap boxing (#1798)

### PR DESCRIPTION
**Pull request by t0k0sh1**
## Summary

- Extend `any` to hold any `enum` value — organic enums (with or without payloads) plus the built-in `Option<T>` / `Result<T, E>` types — via heap-boxing. The 16-byte `RyAny` ABI is preserved by storing a box pointer (`[ArcHeader 16B][descriptor ptr 8B][payload]`) in `any.data[8]` under the new `RyAnyTag::Enum=9`.
- Per-enum-type singleton descriptor (`__ry_enum_desc_<typename>`) carries a discriminant-switched destructor, a variant-wise deep equality function, and the type name. Descriptor cache keys include the full generic parameterization (`Option<int>` ≠ `Option<str>`, `Result<List<int>, str>` ≠ `Result<int, str>`).
- Simple enums (no payload) now flow through `Enum=9` instead of the prior `Int=0` shortcut, so source-level enum identity survives the round-trip: `let c: Color = anyVal` only accepts an `any` actually carrying a `Color`.
- ARC: `wrapInAny` retains the box (and field-wise retains ARC fields in the active variant when the source is an existing enum alias); scope-end cleanup releases via `__ry_arc_dtor_enum_dispatch`. `any == any` matches descriptor pointers first, then dispatches through `desc->eq`. `toStr` / f-string interpolation emits `<TypeName>`. Implicit unwrap is descriptor-pointer-equality only (no record-style subtype chain).

## Implementation notes

- `findEnumLikeTypeNameForBoxing` detection order: organic ADT struct → `Option<T>` / `Result<T, E>` (via reverse map) → simple-enum `i64` carrying `enum_value_type` metadata. The Option/Result branch relies on a new `source_type_name` stamp on the alloca decl so wrapInAny can recover the full generic parameterization across an intermediate `LoadInst` (documented in `.claude/rules/codegen-type-and-metadata.md`).
- Box header offset (−16) matches the collection/record layout; reuses `emitArcGetHeaderFromData`. Box dtor for ADT enums branches on the discriminant and releases only ARC fields in the active variant.
- 10 TDD cycles added covering simple enum round-trip, ADT enum (no payload / int+float / str+List payloads), `Option<T>` / `Result<T, E>` round-trip, implicit unwrap, `List<any>` / `Map<str, any>` with enum elements (regression check for the Tag=Int → Tag=Enum migration), `toStr(<TypeName>)`, and the type-mismatch trap message.

## Test plan

- [x] `./build/ry_tests` — 2394 passed
- [x] `./build/ry test -p` — 195 files, 0 failures
- [x] `./build/ry test tests/spec/any.test.ry` — 168 passed
- [x] Enum-in-collection combinatorial tests (existing) — passed
- [x] ASan + UBSan (`./build-asan/ry test -p` and `./build-asan/ry_tests`) — clean
- [x] TSan (relevant subset: any / arc / option / result / records / combinatorial) — clean
- [x] libFuzzer smoke (200 runs, 4 enum-in-any inputs) — clean
- [x] Background processes / `.serena/` — clean
- [x] Docs updated: `docs/reference/types.md` (Tag=9 row, new "Enums in \`any\`" section, ARC retention semantics)
- [x] CHANGELOG fragment: `changelog.d/1798-any-holds-enum.md`
- [x] Rules updated: `.claude/rules/codegen-arc-cow.md`, `.claude/rules/codegen-type-and-metadata.md`

Closes #1798

## Summary by CodeRabbit

* **New Features**
  * Enums (including Option<T>/Result<T,E>) can be boxed into and unwrapped from any; equality requires exact enum type descriptor match; string/format shows type-name markers; unwraps now enforce exact-type checks and trap on mismatches.
* **Documentation**
  * Reference updated to cover enum-in-any representation, wrap/unwrap semantics, equality, and string formatting.
* **Tests**
  * Added spec and unit tests for enum-in-any wrap/unwrap, equality, reassignment, collections, rendering, and mismatch traps.

[Review Change Stack](https://app.coderabbit.ai/change-stack/t0k0sh1/ry/pull/1807?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

### CODE REVIEW

File: src/codegen_stmt.cpp [hunk 1]
```diff
@@ -654,6 +678,14 @@ void CodeGen::emitVarDecl(const std::string &name,
...
     if (isOptionType(newTy) || isResultType(newTy)) {
         propagateMeta(val, ptr);
+        // #1798: stamp the source-level type name on the alloca so
+        // wrapInAny → findEnumLikeTypeNameForBoxing can recover the full
+        // "Result<List<int>, str>" / "Option<List<int>>" name across the
+        // LoadInst boundary. Without this the fallback path resolves the
+        // inner ptr to "str" (lossy) and the descriptor cache key diverges
+        // from the unwrap-side key.
+        if (annot && !annot->empty())
+            getOrCreateMeta(ptr).source_type_name = *annot;
```
[the remark this answers is not in the document]
Comment (t0k0sh1) [reply src/codegen_stmt.cpp:678]: Fixed in the working tree (commit pending). `src/codegen_stmt.cpp:677-688` now derives `source_type_name` via `inferExprTypeName(value, {}, {})` when `annot` is absent, so an inferred `let o = Some(xs)` / `let r = Ok(xs)` stamps the alloca with the full shaped name (`Option<List<int>>`, `Result<List<int>, str>`) and a subsequent `let a: any = o` routes through the descriptor cache with the same key the unwrap site queries. Reproducer (`xs: List<int> = [1,2,3]; o = Some(xs); a: any = o; backOpt: Option<List<int>> = a`) traps before the fix and prints cleanly after.

File: src/codegen_any.cpp [hunk 2]
```diff
@@ -157,6 +198,21 @@ llvm::Value *CodeGen::unwrapFromAny(llvm::Value *anyVal, llvm::Type *targetTy,
...
+        bool isEnumStruct = !findAdtEnumName(st).empty() ||
+                            isOptionType(st) || isResultType(st);
+        if (isEnumStruct) {
+            return unwrapEnumFromAny(anyVal, st, targetTypeName);
+        }
+    }
+    if (targetTy == i64Ty_ && !targetTypeName.empty() &&
+        isSimpleEnumTypeName(targetTypeName)) {
+        return unwrapEnumFromAny(anyVal, targetTy, targetTypeName);
+    }
```
[the remark this answers is not in the document]
Comment (t0k0sh1) [reply src/codegen_any.cpp:214]: Deferring to a follow-up issue — out of scope for #1798.

Rationale: extending `coerceResultType` to thread an enum-aware target name into the Ok / Err `unwrapFromAny` calls requires:
1. Adding `dstOkTypeName` / `dstErrTypeName` parameters to `coerceResultType` and threading them through ~5 call sites (`emitVarDecl`, reassign at function-local + module-global, return coercion, arg coercion).
2. Relaxing the existing `canAnyHoldType(...)` gate in those call sites so enum-struct destinations (`Option<T>` / ADT) are considered runtime-unwrappable in the Result slot context.
3. Recovering the slot type names from the source-level `Result<X, Y>` annotation, which is not always present at `coerceResultType` time (e.g. function-return coercion where only the LLVM struct type is available).
4. New round-trip tests covering `Result<any, str> -> Result<MyEnum, str>`, `Result<any, E> -> Result<Option<T>, E>`, and the simple-enum-in-Result-Ok corner case.

This is a real corner case but the change is wider than the heap-boxing primitive that #1798 introduces — it widens the static `canAnyHoldType` contract and touches the arg-coercion path that already has its own follow-up backlog. I will file a follow-up issue tracking this so the gap is not lost.

[the remark this answers is not in the document]
Comment (t0k0sh1) [reply src/codegen_any.cpp:214]: Follow-up issue filed: #1808 — `feat(any): extend coerceResultType to enum-aware unwrap of Result payloads (follow-up to #1798)`. Scoped to thread `dstOkTypeName` / `dstErrTypeName` through `coerceResultType` + ~5 call sites, dispatch to `unwrapEnumFromAny` on enum-struct / simple-enum destinations, relax the `canAnyHoldType` gate accordingly, and add `EXPECT_EXIT` tests mirroring `CodeGenTest.AnyEnumUnwrap*Traps` from this PR. Targeted at v0.0.25.
